### PR TITLE
Improve parameter order

### PR DESCRIPTION
**Pull request by HParker**
This resurrects https://github.com/Shopify/yarp/pull/227 (thanks @ccocchi ! I took the liberty of co-authoring you)

and in the second commit uses this newly tracked state to split required params which are after the optional arguments into their own list.

### CODE REVIEW

**Comment by HParker** [comment]
Thank you memcheck!

```
def foo(..., ...)
end
```

was setting the `rest` node twice and therefore leaking the original `rest` node. I don't think I would have found that without valgrind.

File: src/yarp.c [hunk 1]
```diff
@@ -2180,8 +2181,8 @@ yp_parameters_node_location_set(yp_parameters_node_t *params, yp_node_t *param)
 // Append a required parameter to a ParametersNode node.
 static void
 yp_parameters_node_requireds_append(yp_parameters_node_t *params, yp_node_t *param) {
-  yp_parameters_node_location_set(params, param);
-  yp_node_list_append2(&params->requireds, param);
+  yp_parameters_node_location_set(params, (yp_node_t *) param);
```
Comment (kddnewton) [inline src/yarp.c:2184]: It looks like you shouldn't need to cast this since you already have a `yp_node_t *` right?

File: src/yarp.c [hunk 2]
```diff
@@ -2201,8 +2209,8 @@ yp_parameters_node_rest_set(yp_parameters_node_t *params, yp_rest_parameter_node
 // Append a keyword parameter to a ParametersNode node.
 static void
 yp_parameters_node_keywords_append(yp_parameters_node_t *params, yp_node_t *param) {
-  yp_parameters_node_location_set(params, param);
-  yp_node_list_append2(&params->keywords, param);
+  yp_parameters_node_location_set(params, (yp_node_t *) param);
```
Comment (kddnewton) [inline src/yarp.c:2212]: And here

**Review by kddnewton — CHANGES_REQUESTED** [review]
Looking really good! Just a couple of small things.

**Pull request merged**